compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64StringLatin1InflateOp.java

```diff
@@ -0,0 +1,150 @@
+import static org.graalvm.compiler.lir.LIRInstruction.OperandFlag.REG;
+
+@Opcode("AArch64_STRING_INFLATE")
+public final class AArch64StringLatin1InflateOp extends AArch64LIRInstruction {
+    public static final LIRInstructionClass<AArch64StringLatin1InflateOp> TYPE = LIRInstructionClass.create(AArch64StringLatin1InflateOp.class);
+
+    final int chunkByteSize;
+    @Alive({REG}) protected AllocatableValue src;
+    @Alive({REG}) protected AllocatableValue dst;
+    @Alive({REG}) protected AllocatableValue len;
+    @Temp({REG}) protected AllocatableValue temp1;
+    @Temp({REG}) protected AllocatableValue temp2;
+    @Temp({REG}) protected AllocatableValue vectorTemp1;
+    @Temp({REG}) protected AllocatableValue vectorTemp2;
+
+    public AArch64StringLatin1InflateOp(LIRGeneratorTool tool, AllocatableValue src, AllocatableValue dst, AllocatableValue len) {
+        super(TYPE);
+        this.src = src;
+        this.dst = dst;
+        this.len = len;
+        LIRKind archWordKind = LIRKind.value(tool.target().arch.getWordKind());
+        temp1 = tool.newVariable(archWordKind);
+        temp2 = tool.newVariable(archWordKind);
+        LIRKind vectorKind = LIRKind.value(tool.target().arch.getLargestStorableKind(SIMD));
+        vectorTemp1 = tool.newVariable(vectorKind);
+        vectorTemp2 = tool.newVariable(vectorKind);
+        chunkByteSize = 16;
+    }
+
+    @Override
+    public void emitCode(CompilationResultBuilder crb, AArch64MacroAssembler masm) {
+        Label simdImpl = new Label();
+        Label ret = new Label();
+
+        masm.compare(64, asRegister(len), chunkByteSize);
+        masm.branchConditionally(AArch64Assembler.ConditionFlag.GE, simdImpl);
+        emitScalar(masm);
+        masm.jmp(ret);
+
+        masm.bind(simdImpl);
+        emitSIMD(masm, ret);
+
+        masm.bind(ret);
+    }
+
+    private void emitScalar(AArch64MacroAssembler masm) {
+        Register srcAddress = asRegister(temp1);
+        Register destAddress = asRegister(temp2);
+
+        Label loop = new Label();
+
+        try (AArch64MacroAssembler.ScratchRegister scratchReg1 = masm.getScratchRegister(); AArch64MacroAssembler.ScratchRegister scratchReg2 = masm.getScratchRegister()) {
+            Register count = scratchReg1.getRegister();
+            Register currData = scratchReg2.getRegister();
+            masm.mov(64, srcAddress, asRegister(src));
+            masm.mov(64, destAddress, asRegister(dst));
+
+            masm.mov(64, count, asRegister(len));
+
+            masm.align(16);
+            masm.bind(loop);
+            masm.ldr(8, currData, AArch64Address.createImmediateAddress(8, AArch64Address.AddressingMode.IMMEDIATE_POST_INDEXED, srcAddress, 1));
+            masm.str(16, currData, AArch64Address.createImmediateAddress(16, AArch64Address.AddressingMode.IMMEDIATE_POST_INDEXED, destAddress, 2));
+            masm.subs(64, count, count, 1);
+            masm.branchConditionally(AArch64Assembler.ConditionFlag.GE, loop);
```

Comment:
You're off by one here - this should be `ConditionFlag.GT`

Reply:
Done

Reply:
> Our gates are failing on this.
> 
> I've just looked through the code a little, but I see at least one main issue. In the code you are assuming len is 64 bits, but it is a 32 bit value

Cool, now we zero extend the length explicitly from 32 to 64 bits.